### DISCUSSION TITLE
[EMSUSD-1023] meshs complexity in vp2RenderDelegate

### DISCUSSION THREAD

**Discussion by jufrantz** [Autodesk/maya-usd#1809] [Q&A] [answered]
Hello,

Sorry if the topic has already been discussed.

We recently moved to vp2RenderDelegate and users reported that proxyShape.complexity attribute has no longer effect on viewport meshs. They could not get smooth mesh preview anymore, for instance when set dressing to evaluate object contacts.

For test purpose, we tried to compile the dev branch with `HDVP2_ENABLE_GPU_OSD` prototype feature but gave up quickly, some code related to MeshViewportCompute does not compile on our side.

Is there an estimation when complexity attribute would be officially supported ? Also is there a chance that maya 2020 will be supported ? 

Thank you in advance for any information you can provide,
Julien

**Comment by williamkrick** [comment, marked as answer]
Hi @jufrantz,

As you discovered we don't have support for smooth mesh preview in vp2RenderDelegate, and the MeshViewportCompute work is our prototype implementation of the direction we would like to eventually take.

That work relies on Maya API missing in Maya 2020 but present in Maya 2022, which I'm guessing explains your issues compiling with the flag enabled.

Taking the prototype to production ready code is not something we are actively working on at this time.

The major reason for all the complexity here is performance of the SMP code. When proxyShape.complexity is set we smooth every object in the scene, and doing many tiny smoothing operations is slow. We need to run Maya's consolidated world code first to create larger "objects", then do the smoothing, to maximize the performance. Which is complicated and why we have MeshViewportCompute.

There is a simpler solution that you could implement, which would be to add a CPU OSD implementation inside Mesh::Sync() to directly produce individual smoothed objects, which then get consolidated by the normal Maya system. This would work in 2020. The code starting around https://github.com/Autodesk/maya-usd/blob/2efb12e7b4249c4f8e6bf18cf3544c6684303e54/lib/mayaUsd/render/vp2RenderDelegate/meshViewportCompute.cpp#L1113 is the main block of what is necessary to run OSD on the CPU. You'd have to refactor this code to strip away all the consolidation stuff, but it should be possible to have each individual mesh generate a smoothed object. It might even be possible to refactor in a way that MeshViewportCompute and Mesh could share the CPU OSD code, and just have different inputs (consolidated or non-consolidated) depending on the context of the call.

William

**Comment by jufrantz** [comment]
Hello @williamkrick,

Thank you very much for the thorough explanation and for your suggestion.

We will dig more into the code following your idea and maybe come back with a PR. Refining renderItem on CPU during Sync should be acceptable for our first use cases as we wont smooth too big data sets and there will be no mesh animation.

About building with -DHDVP2_ENABLE_GPU_OSD, we tried with 2022 sdk. There were some include problems (it may be because we use USD 21.08) but also these kind of errors :
```
maya-usd/lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp:2415:48: error: 'class pxrInternal_v0_21__pxrReserved__::MeshViewportCompute' has no member named 'level'; did you mean '_level'?
             _meshSharedData->_viewportCompute->level);
```

Best regards,
Julien

**Comment by glevner** [comment]
Just wondering if anyone has made any progress on this issue in the past year and a half? It would be very useful to have a working `complexity` attribute.

I got maya-usd to compile with the `HDVP2_ENABLE_GPU_OSD` macro defined, after pounding on the code a bit. But the resulting plug-in fails to display prims correctly even with `complexity` set to 0...

Geoff

**Comment by dj-mcg** [comment]
Hi @williamkrick & co - I'm jumping on this thread to check in re: status of this request. Is there currently any way to set refinement level for meshes rendered through the VP2 Render Delegate?

**Reply by seando-adsk** [reply]
@dj-mcg Dan in case you weren't aware, William Krick no longer works at Autodesk. @santosd can we try and find the answer to Dan's question.

**Reply by dj-mcg** [reply]
OK, thanks! Sorry about that, I was just responding to the last Autodesk dev I recognized!

**Reply by santosd** [reply]
Hi @dj-mcg , I did some looking around and it doesn't look like we have done any work in regards to mesh complexity support for VP2RenderDelegate. I will do some more checking and testing next week and will get back to you. So far I see that we did some work for nurbs curve complexity support but I didn't find anything for meshes. I will let you know what I find and if I don't see anything, I can log something internally and see if we can take a look.

**Comment by jufrantz** [comment]
Hi @dj-mcg, @glevner, all,

My 2 cents,

We also tried using mayaUsdProxyShape.complexity and I can also confirm that it is still not supported in vp2RenderDelegate. 

In our workflow, we've resorted to using "Edit As Maya Data" and turning on "Smooth Mesh Preview" on pulled maya meshs. While it differs from using "complexity" attribute with pixar proxyShape renderer, it offers the advantage of selectively subdividing meshs in the USD stage.

Support in vp2RenderDelegate would still be highly beneficial, especially if users could activate it individually per mesh, similar to how they would with maya data.

**Reply by dj-mcg** [reply]
I agree, @jufrantz - having this on a per-mesh basis would be really helpful, and would hopefully avoid the performance problems folks encountered when the refinement was being set on the entire scene.

**Reply by santosd** [reply]
Thank you @jufrantz for taking a look, you are correct no work has been done for mesh complexity on the VP2RenderDelegate side. We only display curve complexity. I have logged issue EMSUSD-1023 internally for us to take a look at and see what can be done. I agree that there would be a very useful feature.

**Reply by hiro510** [reply]
Thank you @santosd , you said "We only display curve complexity". What Do you mean?
I interpreted it meant "the current maya-usd plugin can display curves with complexity" but is that right? 
I create a usd file for the simple basis Curve(catmullRom, cubic), then I import it to maya-usd.
The Curve can display without complexity(=Low) with VP2 or Hydra GL. 
What can I control complexity of curve in maya such as usdview?

sample usd file:
#usda 1.0

def BasisCurves "curve"
{
    uniform token basis = "catmullRom"
    int[] curveVertexCounts = [6, 5]
    point3f[] points = [(-50, 0, 0), (-50, 0, 0), (-25, 50, 0), (25, 50, 0), (50, 0, 0), (50, 0, 0), (-50, 50, 0), (-50, 50, 0), (0, 100, 0), (50, 50, 0), (50, 50, 0)]
    color3f[] primvars:displayColor = [(1, 0, 0), (1, 0, 0), (1, 1, 0), (0, 1, 0), (0, 0, 1), (0, 0, 1), (0, 0, 1), (0, 0, 1), (1, 0, 0), (0, 1, 0), (0, 1, 0)] (
        interpolation = "vertex"
    )
    float[] widths = [3.5, 3.5, 2, 1.5, 0.5, 0.1, 3, 3, 0.5, 3, 3] (
        interpolation = "vertex"
    )
}

**Comment by wallworm** [comment]
Our Hydra team is looking into supporting this item in the Hydra viewport. We don't plan on supporting this in VP2.

**Reply by dj-mcg** [reply]
Thanks @wallworm - Is the current plan to eventually deprecate VP2 in favor of the Hydra viewport?

**Reply by wallworm** [reply]
Our viewport team is all-in working on Hydra these days and that is definitely our intent.

**Reply by bjorn-siegert-adsk** [reply]
@dj-mcg you can follow our progress in our git repo here: https://github.com/Autodesk/maya-hydra. It is still early days and there are multiple limitations at the moment which is also the reason why it is a technology preview for now.

**Reply by dj-mcg** [reply]
Thanks folks, glad I more fully understand the direction things are headed. We'd love to start testing our workflows with the Hydra viewport as soon as you think it's in a more complete and stable state.
